### PR TITLE
[libc++] fix back slash as root dir breaks lexically_relative, lexically_proximate and hash_value on Windows

### PR DESCRIPTION
**Pull request by RichardLuo0**
```cpp
#include <filesystem>
#include <iostream>

namespace fs = std::filesystem;
using Path = fs::path;

int main() {
  Path p("C:/a/b");
  Path p2("C:\\a");
  std::cout << p.lexically_proximate(p2) << std::endl;
  std::cout << p.make_preferred().lexically_proximate(p2) << std::endl;
  p = "C:/a";
  p2 = "C:\\a";
  std::cout << (p == p2 ? "true" : "false") << std::endl;
  std::cout << (fs::hash_value(p) == fs::hash_value(p2) ? "true" : "false");
}
```
The above code produces the following result on Windows:
```
"/a\\b"
"b"
true
false
```
lexically_proximate thinks that the root dir is different. However, they are seen as the same thing on windows.
The same issue happens to hash_value(). [The standard](https://timsong-cpp.github.io/cppwp/n4861/fs.class.path#fs.path.nonmember-2) clearly states: If for two paths, p1 == p2 then hash_­value(p1) == hash_­value(p2).

I think we have two solutions:
1. ~~As my PR proposed, make the iterator in the PS_InRootDir state always return "/". If you look at https://en.cppreference.com/w/cpp/filesystem/path/begin, their example has the same behavior. It also prevents other potential issues and saves users time dealing with different root dir slashes.~~
2. Change the implementation of lexically_relative(). Take special care of the PS_InRootDir state. 

Solution 2 is better because it seems it doesn't affect anything else and it is consistent with MSVC behavior (https://godbolt.org/z/GPvx7fc4h)

### CODE REVIEW

File: libcxx/src/filesystem/path.cpp [hunk 1]
```diff
@@ -267,7 +267,7 @@ path path::lexically_relative(const path& base) const {
   // Find the first mismatching element
```
Comment (ldionne) [inline libcxx/src/filesystem/path.cpp:267]: I haven't reviewed this in detail yet, but could you add a test for the property that "`a == b` implies `hash_value(a) == hash_value(b)`"?

File: libcxx/src/filesystem/path.cpp [hunk 2]
```diff
@@ -368,7 +368,8 @@ size_t hash_value(const path& __p) noexcept {
   size_t hash_value = 0;
   hash<string_view_t> hasher;
   while (PP) {
-    hash_value = __hash_combine(hash_value, hasher(*PP));
+    string_view_t Part = PP.inRootDir() ? PATHSTR("/") : *PP;
```
Comment (ldionne) [inline libcxx/src/filesystem/path.cpp:371]: So IIUC you canonicalize `<DRIVE>:/` to just `/` for the purpose of `hash_value`?

**Review by ldionne — COMMENTED** [review]
I feel like I am not sufficiently familiar with Windows paths to make a call here. This seems reasonable but I'd like someone like @mstorsjo to have a look.

**Comment by mstorsjo** [comment]
CC @strega-nil

**Comment by strega-nil** [comment]
I'll take a look today.

File: libcxx/src/filesystem/path.cpp (hunk 1 above)
Comment (RichardLuo0) [reply libcxx/src/filesystem/path.cpp:267]: The original tests already include this: https://github.com/llvm/llvm-project/blob/9f2b4b6c787e0a143d34fcebae8267617fe9e73b/libcxx/test/std/input.output/filesystems/class.path/path.member/path.compare.pass.cpp#L144
But it doesn't cover the different slash as root dir test case, so I add it: https://github.com/llvm/llvm-project/blob/9f2b4b6c787e0a143d34fcebae8267617fe9e73b/libcxx/test/std/input.output/filesystems/class.path/path.member/path.compare.pass.cpp#L81

File: libcxx/src/filesystem/path.cpp (hunk 2 above)
Comment (RichardLuo0) [reply libcxx/src/filesystem/path.cpp:371]: No, the `<DRIVE>:` is [root-name](https://timsong-cpp.github.io/cppwp/n4861/fs.class.path#nt:root-name). Its different from root-directory (which is '/' or '\').
What I did was use '/' to replace the root directory.

**Review by strega-nil — COMMENTED** [review]
This looks reasonable to me - I would want to see testing done on the MSVC-STL to see if the same thing happens over there, but if it does, I'll approve.

**Comment by RichardLuo0** [comment]
> This looks reasonable to me - I would want to see testing done on the MSVC-STL to see if the same thing happens over there, but if it does, I'll approve.

Yes it does. https://godbolt.org/z/GPvx7fc4h

**Review by strega-nil — APPROVED** [review]

**Comment by strega-nil** [comment]
If someone else approves, I'm happy to merge this.

**Pull request merged**